ambry-clustermap/src/test/java/com.github.ambry.clustermap/HelixClusterManagerTest.java

```diff
@@ -0,0 +1,619 @@
+    Utils.writeJsonToFile(zkJson, zkLayoutPath);
+    Utils.writeJsonToFile(testHardwareLayout.getHardwareLayout().toJSONObject(), hardwareLayoutPath);
+    Utils.writeJsonToFile(testPartitionLayout.getPartitionLayout().toJSONObject(), partitionLayoutPath);
+    helixCluster =
+        new MockHelixCluster(clusterNamePrefixInHelix, hardwareLayoutPath, partitionLayoutPath, zkLayoutPath);
+    hostname = "localhost";
+    Properties props = new Properties();
+    props.setProperty("clustermap.host.name", hostname);
+    props.setProperty("clustermap.cluster.name", clusterNamePrefixInHelix + clusterNameStatic);
+    props.setProperty("clustermap.datacenter.name", "DC0");
+    props.setProperty("clustermap.dcs.zk.connect.strings", zkJson.toString(2));
+    clusterMapConfig = new ClusterMapConfig(new VerifiableProperties(props));
+    MockHelixManagerFactory helixManagerFactory = new MockHelixManagerFactory(helixCluster);
+    if (useComposite) {
+      StaticClusterAgentsFactory staticClusterAgentsFactory =
+          new StaticClusterAgentsFactory(clusterMapConfig, hardwareLayoutPath, partitionLayoutPath);
+      metricRegistry = staticClusterAgentsFactory.getMetricRegistry();
+      clusterManager = new CompositeClusterManager(staticClusterAgentsFactory.getClusterMap(),
+          new HelixClusterManager(clusterMapConfig, hostname, helixManagerFactory, metricRegistry));
+    } else {
+      metricRegistry = new MetricRegistry();
+      clusterManager = new HelixClusterManager(clusterMapConfig, hostname, helixManagerFactory, metricRegistry);
+    }
+  }
+
+  /**
+   * Close the cluster managers created.
+   */
+  @After
+  public void after() {
+    if (clusterManager != null) {
+      clusterManager.close();
+    }
+  }
+
+  /**
+   * Test bad instantiation.
+   * @throws Exception
+   */
+  @Test
+  public void badInstantiationTest() throws Exception {
+    // Good test happened in the constructor
+
+    // Bad test
+    Set<ZkInfo> zkInfos = new HashSet<>(dcsToZkInfo.values());
+    zkInfos.iterator().next().port = 0;
+    JSONObject invalidZkJson = constructZkLayoutJSON(zkInfos);
+    Properties props = new Properties();
+    props.setProperty("clustermap.host.name", hostname);
+    props.setProperty("clustermap.cluster.name", clusterNamePrefixInHelix + clusterNameStatic);
+    props.setProperty("clustermap.datacenter.name", "DC0");
+    props.setProperty("clustermap.dcs.zk.connect.strings", invalidZkJson.toString(2));
+    ClusterMapConfig invalidClusterMapConfig = new ClusterMapConfig(new VerifiableProperties(props));
+    metricRegistry = new MetricRegistry();
+    try {
+      new HelixClusterManager(invalidClusterMapConfig, hostname, new MockHelixManagerFactory(helixCluster),
+          metricRegistry);
+      fail("Instantiation should have failed with invalid zk addresses");
+    } catch (IOException e) {
+      assertEquals(true, metricRegistry.getGauges()
```

Comment:
`assertTrue()` requires a cast as the second argument is generic at compile time. Messages are unnecessary in these cases (should be clear from the code that instantiation failure metric was expected to be true but was false actually). Better to keep it less verbose for readability.